src/openrct2/world/Location.hpp

```diff
@@ -129,10 +129,30 @@ struct TileCoordsXYZ
     int32_t x = 0, y = 0, z = 0;
 };
 
+/**
+ * Cardinal directions are represented by the Direction type. It has four
+ * possible values:
+ * 0 is X-decreasing
+ * 1 is Y-increasing
+ * 2 is X-increasing
+ * 3 is Y-decreasing
+ * Direction is not used to model up/down, or diagonal directions.
+ */
+typedef uint8_t Direction;
+
+/**
+ * Given a direction, return the direction that points the other way,
+ * on the same axis.
+ */
+inline Direction direction_reverse(const Direction& dir)
```

Comment:
Why would you choose reference over value here when the type is an integer?

Reply:
I think this can be done as a constexpr (as an equivalent of a define). (cc @IntelOrca )

Reply:
@ZehMatt I don't think it'll make much difference, but I can change it to be pass-by-value if we have any target platforms where the compiler might not handle it as optimally.

@Gymnasiast I don't see how - the output is a function of the input, which is dynamic at runtime?

Reply:
It is recommended to pass small data like that by value otherwise you might end up having code dereferencing memory first, I'm not even sure if the compiler is allowed much margin when it comes to types.

Reply:
I have high confidence in the compiler's ability to do the right thing here, both w.r.t. how it accesses the parameter and w.r.t. optimising the call, but OK, I'll change it just to be safe. We may end up changing it back later if Direction becomes a struct, but we can cross that bridge when it comes to it.

Reply:
```
inline constexpr Direction direction_reverse1(const Direction& dir)
{
    return dir ^ 2;
}
=
unsigned char direction_reverse1(unsigned char const &) PROC           ; direction_reverse1, COMDAT
        movzx   eax, BYTE PTR [rcx]
        xor     al, 2
        ret     0
unsigned char direction_reverse1(unsigned char const &) ENDP           ; direction_reverse1
```
and
```
inline Direction direction_reverse2(const Direction dir)
{
    return dir ^ 2;
}
=
unsigned char direction_reverse2(unsigned char) PROC                        ; direction_reverse2, COMDAT
        xor     cl, 2
        movzx   eax, cl
        ret     0
unsigned char direction_reverse2(unsigned char) ENDP
```

I would be generally careful when it comes to assumptions, I don't believe the compiler can ever go against the type when it comes to references.

Reply:
Whilst this is true, when it is inlined (which it always should be) the reference goes away.